### PR TITLE
[main][feat] transfer collateral between subAccount

### PR DESCRIPTION
**Pull request by chenciao8**

### CODE REVIEW

File: src/services/CrossMarginService.sol [hunk 1]
```diff
@@ -211,6 +218,49 @@ contract CrossMarginService is OwnableUpgradeable, ReentrancyGuardUpgradeable, I
...
+  /// @dev This uses to calculate new trader balance when they tranferring token as collateral.
+  /// @param _primaryAccount Trader's primary address from trader's wallet.
+  /// @param _subAccountIdFrom Trader's Sub-Account Id to withdraw from.
+  /// @param _subAccountIdTo Trader's Sub-Account Id to deposit to.
+  /// @param _token Token that's withdrawn as collateral.
+  /// @param _amount Token withdrawing amount.
+  function transferCollateralSubAccount(
+    address _primaryAccount,
+    uint8 _subAccountIdFrom,
+    uint8 _subAccountIdTo,
```
Comment (kim-hera) [inline src/services/CrossMarginService.sol:231]: Let's discuss this with PO.

```suggestion
    address _primaryAccountFrom,
    address _primaryAccountTo,
    uint8 _subAccountIdFrom,
    uint8 _subAccountIdTo,
```

Comment (chenciao8) [reply src/services/CrossMarginService.sol]: added

**Review by john-xina-p88 — APPROVED** [review]

File: src/services/CrossMarginService.sol [hunk 2]
```diff
@@ -49,6 +49,14 @@ contract CrossMarginService is OwnableUpgradeable, ReentrancyGuardUpgradeable, I
...
     address receiver
   );
+  event LogTransferCollateral(
+    address indexed primaryAccountFrom,
+    address indexed primaryAccountTo,
+    address subAccountFrom,
+    address subAccountTo,
+    address token,
+    uint256 amount
+  );
```
Comment (kim-hera) [inline src/services/CrossMarginService.sol:59]:
```suggestion
  event LogTransferCollateral(
    address indexed primaryAccountFrom,
    uint8 subAccountIdFrom,
    address indexed primaryAccountTo,
    uint8 subAccountIdTo,
    address token,
    uint256 amount
  );
```

File: src/handlers/CrossMarginHandler.sol [hunk 3]
```diff
@@ -283,6 +290,30 @@ contract CrossMarginHandler is OwnableUpgradeable, ReentrancyGuardUpgradeable, I
...
+  ) external nonReentrant onlyAcceptedToken(_token) {
+    if (_amount == 0) revert ICrossMarginHandler_BadAmount();
+    if (_subAccountIdFrom == _subAccountIdTo) revert ICrossMarginHandler_SelfTransfer();
+    // SLOAD
+    CrossMarginService _crossMarginService = CrossMarginService(crossMarginService);
+
+    _crossMarginService.transferCollateral(_primaryAccount, _primaryAccount, _subAccountIdFrom, _subAccountIdTo, _token, _amount);
+
+    emit LogTransferCollateralSubAccount(_primaryAccount, _subAccountIdFrom, _subAccountIdTo, _token, _amount);
+  }
```
Comment (rholterhus) [inline src/handlers/CrossMarginHandler.sol:315]: Was looking through this new PR out of interest, and was wondering if this should be added:

```solidity
require(msg.sender == _primaryAccount);
```

or an equivalent suggestion is to remove `_primaryAccount` and only use `msg.sender`

Comment (kim-hera) [reply src/handlers/CrossMarginHandler.sol:315]: Thanks for the catch!

@chenciao8 Please remove primaryAccount and use msg.sender

Comment (chenciao8) [reply src/handlers/CrossMarginHandler.sol]: Thanks. Fixed.

**Review by kim-hera — APPROVED** [review]

**Pull request merged**